Transcripts - Error and loading states

SergioEstevao opened the pull request:
| 📘 Part of: #1848  |  <!-- project issue number, if applicable -->

This PR adds support for proper styling and localisation of error messages and handling of  the loading state UI.

| Loading | Failed to load | Not supported | Failed to parse |
| - | - | - | - |
|![IMG_0126](https://github.com/user-attachments/assets/c8218ed1-740d-424b-9b79-9269ad8bdab6) | ![IMG_0125](https://github.com/user-attachments/assets/0721e573-9caa-4dc2-b36a-4c83e15cee70) | ![IMG_0124](https://github.com/user-attachments/assets/1b88796b-e5d2-400e-9185-dbec2b2f9fed) | ![IMG_0122](https://github.com/user-attachments/assets/93035002-3f77-4dfc-a701-51108f6b748f) | ![IMG_0121](https://github.com/user-attachments/assets/a0a5639c-9a04-485a-bc6b-eed9da7b7571) |


 
## To test

To test the loading and error scenarios you may want to use a network conditioner, or alternatively just turn off the network connection and change the returned error.

- Start the app
- Ensure that you have the `transcripts` and `newShowNotesEndpoint` FF enabled
- Start playing an episode from a podcast with transcript supports. Ex:  One of this [list](https://lists.pocketcasts.com/b852a088-ccee-4e4b-a3c5-4bb7fd700a02)
- Open the full screen player
- Tap on the Transcript icon

### Test loading indicator

- Check if the loading indicator is show and it's correctly styled  like the above.
- Check that it disappears after the loading finishes

### Test errors

- Disable your network connection
- Play an episode with transcript, because of the cache mechanism you should open an episode that you never open the transcript before
- Open the full screen player
- Tap on transcript
- Check that the error states shows correctly:
  - Error icon shows up 
  - Error message shows up 
  - `Try again` button shows up
  - No Transcript is visible
  - No Search button is visible
  - To test the different error messages edit the line 58 of `TranscriptManager` file and change the `Transcript.failedtoLoad` enum to the different cases available.
- Tap on the `Try again` button
- Check that loading indicator is show
- Enable network connection
- Tap on `Try again`
- Check that loading of transcript is done correctly and error elements dissappear and transcript text shows up correctly

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [ ] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

SergioEstevao committed:
Add error views.

SergioEstevao committed:
Setup text for messages

SergioEstevao committed:
Merge branch 'trunk' into transcripts/error_states

SergioEstevao committed:
Set activity indicator style and color

SergioEstevao committed:
Fix display of error view.

SergioEstevao committed:
Show hide search button depending of loading state.

SergioEstevao committed:
Update error message depending of accessibility text size.

SergioEstevao committed:
Merge branch 'trunk' into transcripts/error_states

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

SergioEstevao committed:
Merge branch 'trunk' into transcripts/error_states

SergioEstevao committed:
Add error icon to error message

SergioEstevao committed:
Add localization support for error messages

SergioEstevao committed:
Fix show hide of transcript

SergioEstevao committed:
Refactor error label variable name

SergioEstevao committed:
Refactor creation of style to single method

SergioEstevao committed:
Set default message

SergioEstevao committed:
Set activity indicator position to centre.

This is needed because the AngularActivityIndicator has it's anchor point on the top left of the component.

leandroalonso commented on podcasts/CommonImages.xcassets/yield_scaled.imageset/Contents.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "images" : [
+    {
+      "filename" : "yield.pdf",
```
We have been using `svg` instead of `pdf`

SergioEstevao replied:
Done here: https://github.com/Automattic/pocket-casts-ios/pull/2007/commits/7963c18148ebd77c836d66ea29efff144e47a82b

leandroalonso commented on podcasts/TranscriptViewController.swift:
```diff
@@ -163,14 +172,55 @@ class TranscriptViewController: PlayerItemViewController {
         return activityIndicatorView
     }()
 
+    private lazy var errorView: UIView = {
```
I wonder if we should extract this to another View?

`TranscriptViewController` is becoming way bigger and by moving the error view to its own view we can keep it simpler. Thoughts?

SergioEstevao replied:
Done here: https://github.com/Automattic/pocket-casts-ios/pull/2007/commits/40a44f81fb4ce21d2301a8ab5c25d914b5ed4599

leandroalonso commented on podcasts/TranscriptViewController.swift:
```diff
@@ -260,6 +310,9 @@ class TranscriptViewController: PlayerItemViewController {
     }
 
     private func loadTranscript() {
+        transcriptView.isHidden = true
+        searchButton.isHidden = true
+        errorView.isHidden = true
         activityIndicatorView.startAnimating()
```
What do you think about moving this to its own method?

SergioEstevao replied:
Done here: https://github.com/Automattic/pocket-casts-ios/pull/2007/commits/f8228ca3732aee5296bf0e7c85228b49f07fa5fd

leandroalonso commented on podcasts/TranscriptViewController.swift:
```diff
@@ -301,32 +359,38 @@ class TranscriptViewController: PlayerItemViewController {
 
     @MainActor
     private func refreshText() {
+        if let errorText = errorLabel.text {
```
I wonder if this should be a new method? `refreshText` was just for transcript before. Now it can be for error OR transcript.

SergioEstevao replied:
Done here: https://github.com/Automattic/pocket-casts-ios/commit/40a44f81fb4ce21d2301a8ab5c25d914b5ed4599

leandroalonso commented on podcasts/TranscriptViewController.swift:
```diff
@@ -301,32 +359,38 @@ class TranscriptViewController: PlayerItemViewController {
-            self.previousRange = nil
-            self.transcript = transcript
-            transcriptView.attributedText = styleText(transcript: transcript)
+        searchButton.isHidden = false
+        errorView.isHidden = true
+        activityIndicatorView.stopAnimating()
+        self.previousRange = nil
```
Same about the previous comment, I believe we can move that to its own method.

SergioEstevao replied:
Done here: https://github.com/Automattic/pocket-casts-ios/pull/2007/commits/f8228ca3732aee5296bf0e7c85228b49f07fa5fd

SergioEstevao committed:
Merge branch 'trunk' into transcripts/error_states

# Conflicts:
#	podcasts/TranscriptViewController.swift

leandroalonso reviewed (changes requested):
Tested all the steps and it works as described!

Left a few comments about extracting the error view so we can simplify `TranscriptViewController`.

SergioEstevao committed:
Refactor error components to separate view.

SergioEstevao committed:
Separate setup of loading states

SergioEstevao committed:
Implement callback for retry

SergioEstevao committed:
Switch to svg asset

SergioEstevao committed:
Merge branch 'trunk' into transcripts/error_states

leandroalonso commented on podcasts/TranscriptViewController.swift:
```diff
@@ -164,14 +173,20 @@ class TranscriptViewController: PlayerItemViewController {
     }()
 
+    private lazy var errorView: TranscriptErrorView = {
+        return TranscriptErrorView {[weak self] in
```
[nit]

```suggestion
       TranscriptErrorView { [weak self] in
```

SergioEstevao committed:
Simplify creation code

Co-authored-by: Leandro Alonso <contato@leandroalonso.com>
